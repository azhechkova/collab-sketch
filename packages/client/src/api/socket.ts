import { io } from 'socket.io-client'
import type { App } from 'vue'

export default {
  install: (app: App<Element>) => {
    const socket = io(import.meta.env.VITE_APP_BASE_URL + '/rooms', { transports: ['websocket'] })
    app.config.globalProperties.$socket = socket
    app.provide('socket', socket)
  }
}
