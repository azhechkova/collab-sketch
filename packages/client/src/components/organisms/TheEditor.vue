<!-- eslint-disable no-case-declarations -->
<script setup lang="ts">
import { inject, ref, watchEffect } from 'vue'
import type { Socket } from 'socket.io-client'

import type { ActionMode, RoomType } from '@/types/index'
import { useEditorStore } from '@/stores/editor'
import { useRoomsStore } from '@/stores/room'
import updateRoomImage from '@/utils/updateRoomImage'
import getFormatClientCoordinates from '@/utils/getFormatClientCoordinates'
import drawLine from '@/utils/drawLine'
import BaseCard from '../atoms/BaseCard.vue'
import { ACTION_MODES } from '@/constants'

const { activeRoom } = defineProps<{
  activeRoom?: string | string[] | null
}>()
const socket = inject('socket') as Socket

const store = useEditorStore()
const roomsStore = useRoomsStore()
const canvasRef = ref<HTMLCanvasElement | null>(null)
const wrapperRef = ref<HTMLDivElement | null>(null)

const setupCanvas = (ctx: CanvasRenderingContext2D) => {
  if (!canvasRef.value || !activeRoom || !wrapperRef.value) return
  store.setCanvas(canvasRef.value)
  store.setWrapper(wrapperRef.value)
  socket.on('draw', ({ prev, point, color, size, roomId }) => {
    if (roomId !== activeRoom) return
    drawLine({ prev, point, color, size }, ctx)
  })

  socket.on('updateRoom', (updatedRoom: RoomType) => {
    updateRoomImage(canvasRef.value, updatedRoom)
  })
}

const onMouseMove = (event: PointerEvent | TouchEvent) => {
  switch (store.actionMode) {
    case ACTION_MODES.draw:
      const { clientX, clientY } = getFormatClientCoordinates(event)

      store.draw(activeRoom, { x: clientX, y: clientY })
      break
    case ACTION_MODES.pan:
      store.onPan(event)
      break
  }
}

const onMouseDown = (e: MouseEvent) => {
  switch (store.actionMode) {
    case ACTION_MODES.draw:
      store.startDrawing(e)
      break

    case ACTION_MODES.pan:
      store.onStartPan()
  }
}

const onMouseUp = () => {
  switch (store.actionMode) {
    case ACTION_MODES.draw:
      store.stopDrawing()
      break
    case ACTION_MODES.pan:
      store.onStopPan()
  }
}

watchEffect(() => {
  const ctx = canvasRef.value?.getContext('2d')
  const activeRoomObj = roomsStore.rooms.find((item) => item._id === activeRoom) || null

  if (ctx && activeRoomObj) {
    updateRoomImage(canvasRef.value, activeRoomObj)
    setupCanvas(ctx)
  }
})

const getCursorStyle = (mode: ActionMode | null) => {
  switch (mode) {
    case 'pan':
      return store.userInteract ? 'grabbing' : 'grab'
    default:
      return 'default'
  }
}
</script>

<template>
  <BaseCard
    :style="{
      cursor: getCursorStyle(store.actionMode)
    }"
    class="relative bg-gray-100 w-[calc(100vw-100px)] h-[calc(100vw-100px)] overflow-hidden"
    @mouseup="onMouseUp"
    @mouseleave="onMouseUp"
    @mousedown="onMouseDown"
    @mousemove="onMouseMove"
  >
    <canvas id="canvasOverlay" class="absolute inset-0" />

    <div ref="wrapperRef" class="absolute bg-white w-[500px] h-[500px]">
      <canvas ref="canvasRef" width="500" height="500" />
    </div>
  </BaseCard>
</template>
