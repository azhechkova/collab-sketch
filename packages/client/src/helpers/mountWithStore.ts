import { createTestingPinia } from '@pinia/testing'
import { mount, VueWrapper, type MountingOptions } from '@vue/test-utils'
import { vi } from 'vitest'
import type { Component, ComponentPropsOptions } from 'vue'

/**
 * Mounts the given Vue component for testing with additional configurations.
 *
 * @param {Component} component - The Vue component to be tested.
 * @param {MountingOptions<ComponentPropsOptions>} options - Mounting options for the component.
 * @returns {VueWrapper} - A mounted test wrapper with createTestingPinia plugin and specified options.
 */

const mountWithStore = (
  component: Component,
  options: MountingOptions<ComponentPropsOptions>
): VueWrapper => {
  return mount(component, {
    global: {
      plugins: [createTestingPinia({ createSpy: vi.fn })]
    },
    ...options
  })
}

export default mountWithStore
