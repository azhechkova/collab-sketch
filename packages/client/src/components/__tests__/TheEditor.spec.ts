import { describe, it, expect } from 'vitest'
import { SocketServerMock } from 'socket.io-mock-ts'

import TheEditor from '../organisms/TheEditor.vue'
import mountWithStore from '@/helpers/mountWithStore'

describe('Editor', () => {
  it('renders properly', () => {
    const socket = new SocketServerMock()

    const wrapper = mountWithStore(TheEditor, {
      props: {
        socket: socket.clientMock
      }
    })
    expect(wrapper.text()).toMatchSnapshot()
  })

  it('renders the canvas element', async () => {
    const socket = new SocketServerMock()

    const wrapper = mountWithStore(TheEditor, {
      props: {
        socket: socket.clientMock
      }
    })

    const canvas = wrapper.find('canvas')
    expect(canvas.exists()).toBe(true)
  })

  it('stops drawing when the mouse button is released', async () => {
    const socket = new SocketServerMock()

    const wrapper = mountWithStore(TheEditor, {
      props: {
        socket: socket.clientMock
      }
    })

    const canvas = wrapper.find('canvas')

    await canvas.trigger('mousedown', { offsetX: 10, offsetY: 10 })
    await canvas.trigger('mousemove', { offsetX: 20, offsetY: 20 })
    await canvas.trigger('mouseup')

    expect(wrapper.vm.$refs.isDrawing).toBeFalsy()
  })
})
