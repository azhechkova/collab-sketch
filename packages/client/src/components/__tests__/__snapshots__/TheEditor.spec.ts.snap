// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Editor > renders properly 1`] = `""`;
